[2d] pass shape/stride during tensor unflatten

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #113547
* #113324
* #113323

as titled, built on top of the work @wz337 enabled, this could save some
runtime CPU time to recreate DTensor parameters with correct
shape/stride, and avoid issues when un-even sharding parameters

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @fduwjj @wz337 @tianyu-l @wconstab @yf225 @kiukchung @d4l3k @lucasllc

wanchaol committed:
[2d] pass shape/stride during tensor unflatten

as titled, built on top of the work @wz337 enabled, this could save some
runtime CPU time to recreate DTensor parameters with correct
shape/stride, and avoid issues when un-even sharding parameters

[ghstack-poisoned]

XilunWu commented on torch/distributed/tensor/parallel/_data_parallel_utils.py:
```diff
@@ -20,5 +20,7 @@ def _unflatten_tensor(tensor: torch.Tensor, spec: DTensorSpec) -> torch.Tensor:
         spec.mesh,
         spec.placements,
         run_check=False,
+        shape=spec.tensor_meta.shape,
```
use `spec.shape`

XilunWu commented on torch/distributed/tensor/parallel/_data_parallel_utils.py:
```diff
@@ -20,5 +20,7 @@ def _unflatten_tensor(tensor: torch.Tensor, spec: DTensorSpec) -> torch.Tensor:
         spec.placements,
         run_check=False,
+        shape=spec.tensor_meta.shape,
+        stride=spec.tensor_meta.stride,
```
add `def stride` just as `shape` and `ndim`?

XilunWu reviewed (commented):
LGTM! Ready to accept when these 2 comments get addressed.

wanchaol committed:
Update on "[2d] pass shape/stride during tensor unflatten"


as titled, built on top of the work wz337 enabled, this could save some
runtime CPU time to recreate DTensor parameters with correct
shape/stride, and avoid issues when un-even sharding parameters

cc H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu

[ghstack-poisoned]

wanchaol committed:
Update on "[2d] pass shape/stride during tensor unflatten"


as titled, built on top of the work wz337 enabled, this could save some
runtime CPU time to recreate DTensor parameters with correct
shape/stride, and avoid issues when un-even sharding parameters

cc H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu

[ghstack-poisoned]

XilunWu reviewed (approved):
LGTM!

wanchaol commented:
@pytorchbot merge

wanchaol commented:
@pytorchbot merge

wanchaol commented:
@pytorchbot merge

wanchaol committed:
Update on "[2d] pass shape/stride during tensor unflatten"


as titled, built on top of the work wz337 enabled, this could save some
runtime CPU time to recreate DTensor parameters with correct
shape/stride, and avoid issues when un-even sharding parameters

cc H-Huang awgu kwen2501 fegin fduwjj wz337 wconstab mrshenli zhaojuanmao rohan-varma kiukchung d4l3k lucasllc XilunWu

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge

wanchaol commented:
@pytorchbot revert -m "broken compile test"

wanchaol commented:
@pytorchbot --help

wanchaol commented:
@pytorchbot revert -m "broken compile test" -c "ghfirst"

wz337 reviewed (approved):
LGTM!

wanchaol commented:
this relanded as a separate PR, closing this
